### ISSUE TITLE
Preparing for Protobuf and JSON Schema support

### ISSUE THREAD

**Issue by dskatz22** [kafkajs/confluent-schema-registry#79]
The confluent schema registry as of version 5.5 supports 2 new schema types, Protobuf and Json Schema, see this [article](https://www.confluent.io/blog/confluent-platform-now-supports-protobuf-json-schema-custom-formats/).  In order to ultimately support these new schema types I would like to make some proposals to the users/contributors of this library. 

Goals
Separate specifics of the Avro schema/serdes from internals of the library. (Cache, Wire Encoder/Decoder, Subject generation)
Maintain backwards compatibility with older Schema Registry deployments.

### Separate Avro from internals
Cache - change the cache to just store the schema string, since other schema types may not be in Json
Wire Encoder/Decoder - change the encoder/decoder to just do the wire encoding/decoding 
Subject Generation - delegate subject generation to specific schema serdes, and future iteration implement the different [strategies ](https://docs.confluent.io/current/schema-registry/serdes-develop/index.html#overview)

All of these changes should push the library to create a Serdes interface that might look something like this.
```
interface Serdes {
  serialize(schema: ConfluentSchema, payload: any) : Buffer
  deserialize(schema: ConfluentSchema, buffer: Buffer) : any 
}
```

### Backwards Compatibility
In version 5.5.0 the schema registry introduces a new field to the various apis called `schemaType`. Reference [here](https://docs.confluent.io/current/schema-registry/develop/api.html#post--subjects-(string-%20subject)-versions).  Since the default is AVRO, and older versions fail when using this field, if the client uses AVRO schema/serdes it should omit the field.

There is a potential for breaking Api changes in this refactor, but users should not need to upgrade their Schema Registry deployment.

Note: I have a rough draft branch of some of these ideas, and want to get feedback before moving forward

**Comment by Nevon** [comment]
Thanks for the detailed issue. I think your plan makes sense. While I'm not the primary author of this library, and am hesitant to take on responsibility for it (hello @erikengervall 👋), I'd be happy to take a look at a pull request and provide feedback.

> There is a potential for breaking Api changes in this refactor,

What potential APIs do you see us having to break backwards compatibility with? I don't have the detailed understanding of the implementation that you most likely do, but I don't immediately see where we would have to break compatibility.

**Comment by dskatz22** [comment]
@Nevon I'll be exploring the Api surface area to try to minimize breaking changes, i.e. constructor defaults to AVRO.  I'll post a draft PR to this issue once I go through that exercise, and make sure there are no test regressions.

**Comment by cristianbriscaru** [comment]
Any progress on adding Json Schema support?

**Comment by yinzara** [comment]
Happy to help, especially on the protobuf side. :)

**Comment by haf** [comment]
This would be great to have! 👍 A workaround for unvalidated messages ought to be

```typescript
message = { ...message, value: message.value.slice(5, message.value.length) }
```

**Comment by malkiz-rapid** [comment]
Hi @dskatz22 and @Nevon ,
What is the status of this issue?

@dskatz22 I have read the code changes in [your branch](https://github.com/kafkajs/confluent-schema-registry/compare/master...dskatz22:serdes-refactor), and I think I understand it.
I would be happy to continue the work you did to add support for Protobuf and JSON Schema.

I just wanted to make sure that this direction is considered acceptable by the library authors/maintainers.

Thanks!

**Comment by Malkiz** [comment]
Hi @dskatz22 and @Nevon ,
What is the status of this issue?

@dskatz22 I have read the code changes in [your branch](https://github.com/kafkajs/confluent-schema-registry/compare/master...dskatz22:serdes-refactor), and I think I understand it.
I would be happy to continue the work you did to add support for Protobuf and JSON Schema.

I just wanted to make sure that this direction is considered acceptable by the library authors/maintainers.

Thanks!

**Comment by Nevon** [comment]
> While I'm not the primary author of this library, and am hesitant to take on responsibility for it (hello @erikengervall 👋), I'd be happy to take a look at a pull request and provide feedback.

Curse my previous self!

I'm currently swamped with being the primary maintainer of KafkaJS along with my day job, so I don't feel like I have the bandwidth to also actively maintain @kafkajs/confluent-schema-registry and cannot commit to helping driving this change through.

In terms of general thoughts, with JSON Schema and Protobuf being officially supported in Confluent schema registry, it makes perfect sense that we should also do so. I can't speak to @dskatz22's implementation, as I don't have the  necessary background on how this is implemented on the registry side and how other clients are working with it.

**Comment by dskatz22** [comment]
Heyall, jumping back in on this.  I'm going to get familiar with this again, since it's been a while.  @Malkiz would you be interested in chatting tomorrow on kafkajs Slack to come up with a plan to make progress on this?

**Comment by Malkiz** [comment]
@dskatz22 sure 👍 I'll ping you on Slack. Thanks!

**Comment by Nick0412** [comment]
What is the status for supporting protobuf? I still receive 
```
ConfluentSchemaRegistryError: not implemented yet
```
every time I try to register a protobuf schema.

**Comment by Nevon** [comment]
Closing this since we shipped Protobuf and JSON Schema support in v2.0.0.
